pywps/inout/outputs.py

```diff
@@ -225,7 +225,7 @@ def _json_data(self, data):
 
                 else:
                     # Otherwise we assume all other formats are unsafe and need to be enclosed in a CDATA tag.
-                    if isinstance(self.data, bytes):
+                    if not isinstance(self.data, bytes):
```

Comment:
decode would be even more incorrect, the output may be a binary file that cannot be decoded as text.
.

Reply:
yes

Reply:
I'm wrong I have to investigate more what is happening in this section of code.

Reply:
It's been a while, but binary files are typically embedded in the response as a base64 string. This bit of code is broken probably because all test cases are detected as having `encoding=base64`. 
We should probably have a test where data is binary, but encoding not base64 to exercise that line.

Reply:
It's look more complex than what I initially thought. I attached the error I got. I investigate it :)

Best regards

[tmp4.txt](https://github.com/geopython/pywps/files/6359261/tmp4.txt)

Reply:
Hello

After more investigation I raise several issues. First point is that in my run case which is a request for a rawdataoutput, the update status try to generate the xml status file, while it is not necessary and not wiched. The second point even if we first the first point we need to fix how complex output data is serialised to json, because it's invalid to encode bytes array. My opinion here is that basically when the mime-type is not one of FORMATS list and the encoding is also not a well known one then we encode in base64. Encoding all unknown data raise a thirds issue, not all the IOHandler variant implement base64.

Another issue that is my personal opinion is that the implementation of IOHandler is insane because it's manipulate the object internals. I'm not against funky implementation, i'm using it often, but here it's look worth than using a composition with ComplexOutput being a proxy to an internal instance of IOHandler created from derivative FileHandler, UrlHandler, DataHandler, etc. on demand.

Thus to fix the issues above I envision to first change the IOHandler implementation/usage by making it a member of complex output instead of having ComplexOutput derivate it. Then implement a conversion to base64 and data for all IOHandler. I still need more think about it, but this is the idea.

What is your opinion ?

Best Regards

Reply:
I fix myself, all IOHandler have base64 conversion available, this make possible to make a simple fix.

Reply:
If you can find a way to simplify the IOHandler without breaking existing code, this would be great. The dynamic subclassing based on attribute assignment `obj.file == ...` converts a generic IOHandler into a FileHandler instance on the fly. 

Previously, the File, Data, Stream and Url handling logic was all sprinkled into the IOHandler class and made it hard to maintain and extend. I've tried to clarify the logic by encapsulating it into different classes, but I had to rig this exotic mechanism to preserve backward compatibility. 

One issue is that when a Process is created, we don't know if the outputs are going to be defined as data, file, url or a stream. This is not part of the process definition, but rather the implementation of the handler.

Reply:
Hello huard,

I will have a look to this and try to simplify the logic a little bit.

Best regards